### ISSUE TITLE
Correct usage of nl_cache_mngr

### ISSUE THREAD

**Issue by davidchappelle** [thom311/libnl#71]
I am trying to utilize the cache manager to track changes to the network interfaces and their corresponding ip addresses. However, I am running into some misunderstandings in terms of how the cache manager is supposed to work. Most importantly, the change callback is only called with the newly modified nl_link object. The problem is that there are a large number of attributes that can change on an interface and the old nl_link object is not provided. So there is no way to actually know what has changed without performing another layer of caching. Is this how it is intended to work? I am supposed to build a thin caching layer on top of the cache manager? How else would someone be able to generate events to an application based on the cache manager?

**Comment by thom311** [comment]
Yes, you don't know what changed.

I think most applications wouldn't care about that distinction. Can you not just do a full reload of the reported item on every event? The cache-manager is a simple solution that works for many use cases.


-- how about...

Currently there are the functions nl_cache_include() and nl_cache_resync() which eventually call cache_include(). The latter doesn't give you the flexibility you need. nl_cache could be extended to provided two more powerful alternatives to nl_cache_include() and nl_cache_resync() with a more expressive change_func_t callback.

But first of all, it's not clear how a better change_func_t would be. Internally, libnl has change flags that are returned by nl_object_diff(). But they are not public API. maybe the should made public and passed to the callback. That would be a larger change...


How about adding new functions (and keep the current ones for backward-compat):

    int nl_cache_resync_v2(struct nl_sock *sk, struct nl_cache *cache,
                           change_func_t_v2 change_cb, void *data);
    int nl_cache_include_v2(struct nl_cache *cache, struct nl_object *obj,
                            change_func_t_v2 change_cb, void *data);
    typedef int (*change_func_t_v2)(struct nl_cache *cache, int action,
             struct nl_object *obj, void *);

Basically, instead of calling cache_include(), it would call change_func_t_v2. That callback can reimplement cache_include() however he wants. Of if it returns 0 (not-handled-by-callback), nl_cache() would proceed with what it currently does via cache_include().

**Comment by davidchappelle** [comment]
If I want to generate notifications for network interface up/down and ip address change it is important to provide both objects. I found this article on just that very thing. Seems to be inline with what you are proposing above except the author hacks the API to achieve it.

    http://carnivore.it/2012/11/28/libnl_-_example_code

In this code the change callback takes both the old and the new object.

**Comment by thom311** [comment]
this is now fixed by recent patch from Tobias Jungel. Closing as fixed.
